src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt

```diff
@@ -1,192 +1,105 @@
-  private var codyOnboardingGuidancePanel: CodyOnboardingGuidancePanel? = null
-  private val chatMessageHistory = CodyChatMessageHistory(CHAT_MESSAGE_HISTORY_CAPACITY)
-
-  init {
-    // Tabs
-    val contentPanel = JPanel()
-    tabbedPane.insertTab(
-        /* title = */ "Chat",
-        /* icon = */ null,
-        /* component = */ contentPanel,
-        /* tip = */ null,
-        CHAT_TAB_INDEX)
 
-    tabbedPane.insertTab(
-        /* title = */ "Commands",
-        /* icon = */ null,
-        /* component = */ commandsPanel,
-        /* tip = */ null,
-        RECIPES_TAB_INDEX)
-    subscriptionPanel = SubscriptionTabPanel()
-    // Chat panel
-    messagesPanel.layout = VerticalFlowLayout(VerticalFlowLayout.TOP, 0, 0, true, true)
-    val chatPanel = ChatScrollPane(messagesPanel)
-
-    // Controls panel
-    sendButton = createSendButton()
-    promptPanel =
-        PromptPanel(
-            chatMessageHistory,
-            ::sendChatMessage,
-            sendButton,
-            isGenerating = stopGeneratingButton::isVisible)
-    val stopGeneratingButtonPanel = JPanel(FlowLayout(FlowLayout.CENTER, 0, 5))
-    stopGeneratingButtonPanel.minimumSize = Dimension(Short.MAX_VALUE.toInt(), 0)
-    stopGeneratingButton.addActionListener {
-      inProgressChat.abort()
-      stopGeneratingButton.isVisible = false
-      sendButton.isEnabled = promptPanel.textArea.text.isNotBlank()
-      ensureBlinkingCursorIsNotDisplayed()
-      commandsPanel.enableAllButtons()
-    }
+  private val commandsPanel =
+      CommandsTabPanel(project) { commandId: CommandId ->
+        addChatSession(AgentChatSession.createFromCommand(project, commandId))
+      }
 
-    stopGeneratingButton.isVisible = false
-    stopGeneratingButtonPanel.add(stopGeneratingButton)
-    stopGeneratingButtonPanel.isOpaque = false
-    contextView = EnhancedContextPanel(project)
-    val lowerPanel = LowerPanel(stopGeneratingButtonPanel, promptPanel, contextView)
+  private val subscriptionPanel = SubscriptionTabPanel()
 
-    // Main content panel
-    contentPanel.layout = BorderLayout(0, 0)
-    contentPanel.border = BorderFactory.createEmptyBorder(0, 0, 10, 0)
-    contentPanel.add(chatPanel, BorderLayout.CENTER)
-    contentPanel.add(lowerPanel, BorderLayout.SOUTH)
-    tabbedPane.addChangeListener { focusPromptInput() }
-    val singInWithSourcegraphPanel = SignInWithSourcegraphPanel(project)
-    allContentPanel.add(tabbedPane, "tabbedPane", CHAT_PANEL_INDEX)
-    allContentPanel.add(
-        singInWithSourcegraphPanel, SING_IN_WITH_SOURCEGRAPH_PANEL, SIGN_IN_PANEL_INDEX)
-    allContentLayout.show(allContentPanel, SING_IN_WITH_SOURCEGRAPH_PANEL)
-    refreshPanelsVisibility()
+  init {
+    CodyAgentService.getInstance(project).onStartup { agent ->
+      agent.client.onNewMessage = Consumer { params ->
+        AgentChatSession.getSession(params.id)?.receiveMessage(params.message)
+      }
 
-    addWelcomeMessage()
+      AgentChatSession.restoreAllSessions(agent)
```

Comment:
Why do we call /restore for all sessions during startup? 

I attached a debugger here (on `AgentChatSession.restoreAllSessions(agent)`) solving another issue and this seems a bit odd to me. This constructor is invoked during the toolbar creation. When it's executed there's one session without messages. So an empty, initial session is restored.

Reply:
Ha! FYI I commented out this method along with its usage and I don't see any difference (including working /restore). :smile:

Reply:
Discussed offline, but for other readers:
> So an empty, initial session is restored.
That is indeed a side effect, although harmless.
Purpose of this block is to be ran on **every agent restart** and restore all connections/sessionsId/chat history.
So it's crucial and cannot be removed :)